docs/confirmations.md

```diff
@@ -0,0 +1,253 @@
+| opts.requestState | Additional mutable data for the request.<br>Must be a JSON compatible object.<br>Can be updated using the `ApprovalController.updateRequestState` action. | `{ status: 'pending' }` |
+| shouldShowRequest | A boolean indicating whether the popup should be displayed. | `true` |
+
+#### Example
+
+```
+await this.messagingSystem.call(
+  'ApprovalController:addRequest',
+  {
+    id,
+    origin,
+    type,
+    requestData,
+  },
+  true,
+);
+```
+
+### 3. Update Approval Request
+
+If you wish to provide additional state to the confirmation while it is visible, send an `updateRequestState` message to the `ApprovalController`.
+
+This requires you to have provided the `id` when creating the approval request, so it can be passed to the update message.
+
+The available message arguments are:
+
+| Name | Description | Example Value |
+| -- | -- | -- |
+| opts.id | The ID of the approval request to update. | `"f81f5c8a-33bb-4f31-a4e2-52f8b94c393b"` |
+| opts.requestState | The updated mutable data for the request.<br>Must be a JSON compatible object. | `{ status: 'pending' }` |
+
+#### Example
+
+```
+await this.messagingSystem.call(
+  'ApprovalController:updateRequestState',
+  {
+    id,
+    requestState: { counter },
+  },
+);
+```
+
+## Frontend
+
+### 1. Create Template File
+
+The `ConfirmationPage` component is already configured to display any approval requests generated by the `ApprovalController` and the associated `pendingApprovals` state.
+
+In order to configure how the resulting confirmation is rendered, an **Approval Template** is required.
+
+Create a new JavaScript file in `ui/pages/confirmation/templates` with the name matching the `type` used in the background approval request.
+
+### 2. Update Approval Templates
+
+Add your imported file to the `APPROVAL_TEMPLATES` constant in:
+[ui/pages/confirmation/templates/index.js](../ui/pages/confirmation/templates/index.js)
+
+### 3. Define Values
+
+Inside the template file, define a `getValues` function that returns an object with the following properties:
+
+| Name | Description | Example Value |
+| -- | -- | -- |
+| content | An array of objects defining the components to be rendered in the confirmation.<br>Processed by the [MetaMaskTemplateRenderer](../ui/components/app/metamask-template-renderer/metamask-template-renderer.js).  | See example below. |
+| onSubmit | A callback to execute when the user approves the confirmation. | `actions.resolvePendingApproval(...)` |
+| onCancel | A callback to execute when the user rejects the confirmation. | `actions.rejectPendingApproval(...)` |
+| submitText | Text shown for the accept button. | `t('approveButtonText')` |
+| cancelText | Text shown on the reject button. | `t('cancel')` |
+| loadingText | Text shown while waiting for the onSubmit callback to complete. | `t('addingCustomNetwork')` |
+| networkDisplay | A boolean indicating whether to show the current network at the top of the confirmation. | `true` |
+
```

Comment:
Hey @matthewwalsh0 : above frontend code is specific to snaps. It is not used by extension confirmation pages. If we are adding a new confirmation for instance new signature request to the extension, what should be the approach ?

Reply:
Discussed offline, used by snaps but also the extension, currently for adding and switching networks.